File: android/app/src/main/java/com/budgetmanager/data/repository/InMemoryBudgetRepository.kt
package com.budgetmanager.data.repository

import com.budgetmanager.domain.entity.Budget
import com.budgetmanager.domain.entity.DomainError
import com.budgetmanager.domain.repository.BudgetRepository
import java.time.LocalDateTime

class InMemoryBudgetRepository : BudgetRepository {
    private val budgets = mutableListOf<Budget>()
    private var nextId: Long = 1

    override suspend fun getAll(): List<Budget> {
        return budgets.sortedByDescending { it.period }
    }

    override suspend fun getById(id: Long): Budget {
        return budgets.find { it.id == id } ?: throw DomainError.NotFound
    }

    override suspend fun getByPocketId(pocketId: Long): List<Budget> {
        return budgets
            .filter { it.pocketId == pocketId }
            .sortedByDescending { it.period }
    }

    override suspend fun getByPeriod(period: String): List<Budget> {
        return budgets
            .filter { it.period == period }
            .sortedBy { it.name }
    }

    override suspend fun add(budget: Budget): Budget {
        val newBudget = budget.copy(id = nextId++)
        budgets.add(newBudget)
        return newBudget
    }

    override suspend fun update(budget: Budget) {
        val index = budgets.indexOfFirst { it.id == budget.id }
        if (index == -1) throw DomainError.NotFound

        budgets[index] = budget.copy(updatedAt = LocalDateTime.now())
    }

    override suspend fun updateSpentAmount(id: Long, amount: Double) {
        val index = budgets.indexOfFirst { it.id == id }
        if (index == -1) throw DomainError.NotFound

        val budget = budgets[index]
        budgets[index] = budget.copy(
            spentAmount = budget.spentAmount + amount,
            updatedAt = LocalDateTime.now()
        )
    }

    override suspend fun delete(id: Long) {
        if (!budgets.removeIf { it.id == id }) {
            throw DomainError.NotFound
        }
    }

    override fun hasBudgets(pocketId: Long): Boolean {
        return budgets.any { it.pocketId == pocketId }
    }
}

File: iOS/Sources/Data/Repositories/InMemoryBudgetRepository.swift
import Foundation

class InMemoryBudgetRepository: BudgetRepositoryProtocol {
    private var budgets: [Budget] = []
    private var nextId: Int64 = 1

    func fetchAll() async throws -> [Budget] {
        return budgets.sorted { $0.period > $1.period }
    }

    func fetch(id: Int64) async throws -> Budget {
        guard let budget = budgets.first(where: { $0.id == id }) else {
            throw DomainError.notFound
        }
        return budget
    }

    func fetchByPocketId(_ pocketId: Int64) async throws -> [Budget] {
        return budgets
            .filter { $0.pocketId == pocketId }
            .sorted { $0.period > $1.period }
    }

    func fetchByPeriod(_ period: String) async throws -> [Budget] {
        return budgets
            .filter { $0.period == period }
            .sorted { $0.name < $1.name }
    }

    func add(_ budget: Budget) async throws -> Budget {
        let newBudget = Budget(
            id: nextId,
            name: budget.name,
            description: budget.description,
            pocketId: budget.pocketId,
            allocatedAmount: budget.allocatedAmount,
            spentAmount: budget.spentAmount,
            period: budget.period,
            createdAt: budget.createdAt,
            updatedAt: budget.updatedAt
        )
        nextId += 1
        budgets.append(newBudget)
        return newBudget
    }

    func update(_ budget: Budget) async throws {
        guard let index = budgets.firstIndex(where: { $0.id == budget.id }) else {
            throw DomainError.notFound
        }
        var updated = budget
        updated.updatedAt = Date()
        budgets[index] = updated
    }

    func updateSpentAmount(id: Int64, amount: Double) async throws {
        guard let index = budgets.firstIndex(where: { $0.id == id }) else {
            throw DomainError.notFound
        }
        budgets[index].spentAmount += amount
        budgets[index].updatedAt = Date()
    }

    func delete(id: Int64) async throws {
        guard budgets.contains(where: { $0.id == id }) else {
            throw DomainError.notFound
        }
        budgets.removeAll { $0.id == id }
    }

    func hasBudgets(forPocketId pocketId: Int64) -> Bool {
        budgets.contains { $0.pocketId == pocketId }
    }
}
